[usdt][torch] Sample dispatch operator integration

RihamSelim opened the pull request:
Summary:
By default the instruction at the USDT is nop, when the tracepoint is attached (e.g. through bpftrace) the code inside the semaphore check is executed. Thus there should be no performance impact as long as the USDT is not attached from the tracepoint execution code itself, however the semaphore check itself `TORCH_SDT_IS_ENABLED` will incur the cost of a `read_global_volatile` operation.

https://github.com/dtrace4linux/linux/blob/master/doc/usdt.html for more info

Test Plan:
```
buck2  build  mode/opt caffe2/torch/fb/observers:strobelight_observer_runner --show-full-output
```

```
/data/users/rihams/fbsource/buck-out/v2/gen/fbcode/0bc8cf217a8cf352/caffe2/torch/fb/observers/__strobelight_observer_runner__/strobelight_observer_runner
```

```
sudo bpftrace -e 'usdt:/data/users/rihams/fbsource/buck-out/v2/gen/fbcode/6081734815403318/caffe2/torch/fb/observers/__strobelight_observer_runner__/strobelight_observer_runner:pytorch:operator_* { printf("%s --> %s\n", probe, str(arg0)); }' -v

usdt:<path>strobelight_observer_runner:pytorch:operator_start --> aten::empty_strided
usdt:<path>strobelight_observer_runner:pytorch:operator_end --> aten::empty_strided
usdt:<path>strobelight_observer_runner:pytorch:operator_end --> aten::empty_like
usdt:<path>strobelight_observer_runner:pytorch:operator_start --> aten::fill_
usdt:<path>strobelight_observer_runner:pytorch:operator_end --> aten::fill_
usdt:<path>strobelight_observer_runner:pytorch:operator_end --> aten::ones_like
usdt:<path>strobelight_observer_runner:pytorch:operator_start --> aten::mul
usdt:<path>strobelight_observer_runner:pytorch:operator_end --> aten::mul
usdt:<path>strobelight_observer_runner:pytorch:operator_start --> aten::add
usdt:<path>strobelight_observer_runner:pytorch:operator_end --> aten::add
usdt:<path>strobelight_observer_runner:pytorch:operator_start --> aten::detach
usdt:<path>strobelight_observer_runner:pytorch:operator_end --> aten::detach
usdt:<path>strobelight_observer_runner:pytorch:operator_start --> aten::randn
usdt:<path>strobelight_observer_runner:pytorch:operator_start --> aten::empty
usdt:<path>strobelight_observer_runner:pytorch:operator_end --> aten::empty
usdt:<path>strobelight_observer_runner:pytorch:operator_start --> aten::normal_
usdt:<path>strobelight_observer_runner:pytorch:operator_end --> aten::normal_
usdt:<path>strobelight_observer_runner:pytorch:operator_end --> aten::randn
usdt:<path>strobelight_observer_runner:pytorch:operator_start --> aten::to
usdt:<path>strobelight_observer_runner:pytorch:operator_start --> aten::_to_copy
usdt:<path>strobelight_observer_runner:pytorch:operator_start --> aten::empty_strided
usdt:<path>strobelight_observer_runner:pytorch:operator_end --> aten::empty_strided

```

Differential Revision: D44636587

malfet commented on aten/src/ATen/core/dispatch/Dispatcher.h:
```diff
@@ -307,6 +335,9 @@ class TORCH_API Dispatcher final {
   static int64_t sequenceNumberForRunningRecordFunction(DispatchKey dispatchKey);
   static void runRecordFunction(at::RecordFunction& guard, at::RecordFunction::schema_ref_t schema_ref, DispatchKey dispatchKey);
   static void runRecordFunction(at::RecordFunction& guard, at::RecordFunction::schema_ref_t schema_ref, DispatchKey dispatchKey, c10::ArrayRef<const c10::IValue> args);
+  static bool profilingOperatorEvents();
+  static void fireOpStartUSDT(at::RecordFunction::schema_ref_t schema_ref);
+  static void fireOpEndUSDT(at::RecordFunction::schema_ref_t schema_ref);
```
Don't you want to wrap it with `#ifdef FBCODE_CAFFE2`

malfet commented on aten/src/ATen/core/dispatch/Dispatcher.h:
```diff
@@ -670,7 +701,19 @@ C10_ALWAYS_INLINE_UNLESS_MOBILE Return Dispatcher::call(const TypedOperatorHandl
+
+#ifdef FBCODE_CAFFE2
+  if(profilingOperatorEvents()) {
+    auto& schema = op.schema();
+    auto schema_ref = std::reference_wrapper<const FunctionSchema>(schema);
+    EventRAII ev(fireOpStartUSDT, fireOpEndUSDT, schema_ref);
+    return kernel.template call<Return, Args...>(op, dispatchKeySet, std::forward<Args>(args)...);
```
How about moving FireOpRAII into the template itself and avoid exposing it globally and two unnecessary binds? (one can even use `https://en.cppreference.com/w/cpp/memory/unique_ptr` with custom deleter, though this would look a bit ugly)
```suggestion
    struct FireOpRAII {
       FireOpRAII(at::RecordFunction::schema_ref_t sr): sr_(sr) { 
           fireOpStartUSDT(sr);
        }
       ~FireOpRAII() { fireOpEndUSDT(sr_); }
       at::RecordFunction::schema_ref_t sr;
    } event(op.schema());
    return kernel.template call<Return, Args...>(op, dispatchKeySet, std::forward<Args>(args)...);
```

RihamSelim committed:
[usdt][torch] Sample dispatch operator integration (#115593)

Summary:

Add a USDT for start and end of PyTorch operator.
By default the instruction at the USDT is nop, when the tracepoint is attached (e.g. through bpftrace) the code inside the semaphore check is executed. Thus there should be no performance impact as long as the USDT is not attached.
https://github.com/dtrace4linux/linux/blob/master/doc/usdt.html for more info

Test Plan:
```
buck2  build  mode/opt caffe2/torch/fb/observers:strobelight_observer_runner --show-full-output
```

```
/data/users/rihams/fbsource/buck-out/v2/gen/fbcode/0bc8cf217a8cf352/caffe2/torch/fb/observers/__strobelight_observer_runner__/strobelight_observer_runner
```

```
sudo bpftrace -e 'usdt:/data/users/rihams/fbsource/buck-out/v2/gen/fbcode/6081734815403318/caffe2/torch/fb/observers/__strobelight_observer_runner__/strobelight_observer_runner:pytorch:operator_* { printf("%s --> %s\n", probe, str(arg0)); }' -v

usdt:<path>strobelight_observer_runner:pytorch:operator_start --> aten::empty_strided
usdt:<path>strobelight_observer_runner:pytorch:operator_end --> aten::empty_strided
usdt:<path>strobelight_observer_runner:pytorch:operator_end --> aten::empty_like
usdt:<path>strobelight_observer_runner:pytorch:operator_start --> aten::fill_
usdt:<path>strobelight_observer_runner:pytorch:operator_end --> aten::fill_
usdt:<path>strobelight_observer_runner:pytorch:operator_end --> aten::ones_like
usdt:<path>strobelight_observer_runner:pytorch:operator_start --> aten::mul
usdt:<path>strobelight_observer_runner:pytorch:operator_end --> aten::mul
usdt:<path>strobelight_observer_runner:pytorch:operator_start --> aten::add
usdt:<path>strobelight_observer_runner:pytorch:operator_end --> aten::add
usdt:<path>strobelight_observer_runner:pytorch:operator_start --> aten::detach
usdt:<path>strobelight_observer_runner:pytorch:operator_end --> aten::detach
usdt:<path>strobelight_observer_runner:pytorch:operator_start --> aten::randn
usdt:<path>strobelight_observer_runner:pytorch:operator_start --> aten::empty
usdt:<path>strobelight_observer_runner:pytorch:operator_end --> aten::empty
usdt:<path>strobelight_observer_runner:pytorch:operator_start --> aten::normal_
usdt:<path>strobelight_observer_runner:pytorch:operator_end --> aten::normal_
usdt:<path>strobelight_observer_runner:pytorch:operator_end --> aten::randn
usdt:<path>strobelight_observer_runner:pytorch:operator_start --> aten::to
usdt:<path>strobelight_observer_runner:pytorch:operator_start --> aten::_to_copy
usdt:<path>strobelight_observer_runner:pytorch:operator_start --> aten::empty_strided
usdt:<path>strobelight_observer_runner:pytorch:operator_end --> aten::empty_strided

```

Differential Revision: D44636587

malfet commented:
@pytorchbot merge
